cs/src/core/ClientSession/ClientSession.cs

```diff
@@ -747,13 +747,53 @@ public void ConcurrentReader(ref Key key, ref Input input, ref Value value, ref
 
             public bool ConcurrentWriter(ref Key key, ref Value src, ref Value dst, long address)
             {
-                return _clientSession.functions.ConcurrentWriter(ref key, ref src, ref dst);
+                if (!_clientSession.fht.SupportsMutableIndexes)
+                    return _clientSession.functions.ConcurrentWriter(ref key, ref src, ref dst);
+
+                ref RecordInfo recordInfo = ref _clientSession.fht.RecordAccessor.SpinLockRecordInfo(address);
+                try
+                {
+                    if (!recordInfo.Tombstone && _clientSession.functions.ConcurrentWriter(ref key, ref src, ref dst))
+                    {
+                        // KeyIndexes do not need notification of in-place updates because the key does not change.
+                        if (_clientSession.fht.SecondaryIndexBroker.MutableValueIndexCount > 0)
```

Comment:
Keys in FASTER are immutable once added to the log, so we do not need a lock when only accessing keys.